Wrappers for Mantid workspace classes

MatthewBowles opened the pull request:
Added workspace wrappers according to description in #151.

**To test:**
Run unit tests

Fixes #151 

@mducle @martyngigg

martyngigg commented on mslice/workspace/workspace.py:
```diff
@@ -0,0 +1,77 @@
+    def __init__(self, matrix_workspace):
+        self.inner_workspace = matrix_workspace
+
+    def get_coordinates(self):
```
Can we have some method documentation here and on the other methods, please? See https://www.python.org/dev/peps/pep-0257/

MatthewBowles commented on mslice/tests/workspace_test.py:
```diff
@@ -0,0 +1,217 @@
+import unittest
```
`PixelWorkspaceTest` definitely belongs in its own file, but I'm unsure about the others.

The tests for `Workspace` and `HistogramWorkspace` are basically the same, so they inherit from `BaseWorkspaceTest` (which doesn't actually have any tests on its own, just provides common implementation to the two subclasses). We might want to keep these in the same file since they're so closely related?

martyngigg replied:
@MatthewBowles I think it would still be good to split them up. If I'm looking for the `HistogramWorkspace` test then it's not that obvious that it's with the workspace one. Can you just import the `BaseWorkspaceTest` class in the `Histogram` test?

MatthewBowles replied:
Good point, will do.

martyngigg commented on mslice/workspace/base.py:
```diff
@@ -0,0 +1,48 @@
+class WorkspaceBase(object):
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, mantid_workspace):
```
I think we would be better leaving `__init__` out completely as this is now an interface so it would be better that it has no implementation.

mducle commented:
@martyngigg @MatthewBowles It looks fine to me. I've just run a quick check and everything seems to work as advertised (except when I purposefully try to break it by constructing things from a non-workspace type (or the wrong workspace type).

martyngigg commented on mslice/workspace/workspace.py:
```diff
@@ -0,0 +1,103 @@
+    """workspace wrapper for MatrixWorkspace"""
+
+    def __init__(self, mantid_workspace):
+        self._raw_ws = mantid_workspace
```
Thinking about this I wonder if we might be better with a `mixin-style` hierarchy (https://www.ianlewis.org/en/mixins-and-python). This would allows us to decouple the types in the inheritance hierarchy so that, for example, `PixelWorkspace` and `Workspace` do not inherit each other but both inherit `WorkspaceBase` and then we mixin other classes to add the functionality.

MatthewBowles commented:
@martyngigg I've implemented the mixin classes. I tried to keep them loosely-coupled and added a `rewrap` method to help with this. Changed the `__pow__` methods to be more efficient as well.
The CI build is giving 404 errors for some reason though.

martyngigg commented:
@MatthewBowles There was an issue with the nightly deployment of the mantid build. I've fixed that and restarted it.

martyngigg commented on mslice/workspace/histo_mixin.py:
```diff
@@ -0,0 +1,51 @@
+            raise RuntimeError("List or array must have same number of elements as an axis of the workspace")
+
+    def __pow__(self, exponent):
+        pow_alg = AlgorithmManager.createUnmanaged('PowerMD')
```
Can we do the same with `CreateMDHistoWorkspace` and `ReplicateMD` above? Maybe a utility function for running an  algorithm in this way that accepts `(name, *args, **kwargs)`?

MatthewBowles replied:
@martyngigg This seems like something that would be useful in Mantid also - this is a common pattern for running algorithms but it's quite verbose. Maybe it would be better to implement something in Mantid that Mslice could import.

martyngigg replied:
@MatthewBowles Agreed. @gvardany is working on something now where you would be able to do `ws = PowerMD(..., RunAsChild=True)` that would achieve the same result. I don't think we should wait for that.

martyngigg commented:
@MatthewBowles The builds look to be fixed now.

martyngigg commented:
@MatthewBowles Could you make those minor changes to the way the algorithms are run? We can have a future maintenance task once Mantid has a sensible generic implemetation for this.

MatthewBowles commented:
@martyngigg I've made the changes but the build keeps giving seg faults?
